### PR TITLE
Allow a custom paging behavior to become app default #284

### PR DESCRIPTION
**Pull request by weaselmetal**
Here's my PR

### CODE REVIEW

File: crnk-core/src/main/java/io/crnk/core/engine/internal/information/resource/DefaultResourceInformationProvider.java [hunk 1]
```diff
@@ -82,9 +83,21 @@ public ResourceInformation build(Class<?> resourceClass, boolean allowNonResourc
...
+							.filter(it -> pagingBehaviorType.isInstance(it))
+							.findFirst();
+
+			if (!optPagingBehavior.isPresent()) {
+				throw new IllegalStateException("paging behavior not found: " + pagingBehaviorType);
+			}
+
+		} else {
+			optPagingBehavior = pagingBehaviors.values().stream().findFirst();
 		}
```
Comment (weaselmetal) [inline crnk-core/src/main/java/io/crnk/core/engine/internal/information/resource/DefaultResourceInformationProvider.java]: Those lines above would pretty much be the same, even if a delegating `DefaultPagingBehavior` was used (except that I'd have to instantiate a `DefaultPagingBehavior` that delegates to the actual implementation).
I decided against that, as it only makes it a little harder to figure out which behavior is used. @remmeier

**Comment by weaselmetal** [comment]
please let me know what you think about that PR @remmeier

**Comment by remmeier** [comment]
thx, what is the state regarding the forwarding vs null/replacing behavior? is forwarding still the way to go I think? Or is there an issue? If possible I would really like to simplify the crnk engine/setup part.

**Comment by weaselmetal** [comment]
Take note of the review I started @remmeier. It's possible to delegate from a default behavior to a concrete behavior, but the code I commented on wouldn't change much. So I'd go for the replace idea, as you'd immediately see the concrete behavior.

Anyway, my last commit broke tests, and I wonder what the test expectation is.
Right now, in the test, a Custom paging behavior is registered against the framework.
But the test seems to expect, that a registered custom behavior is NOT used for classes, that don't specify any paging behavior. Rather, the default OffsetLimit paging behavior seems to be expected for such classes. But that is in conflict with what I was trying to achieve. So I guess way to go would be to adapt the tests?

**Comment by remmeier** [comment]
I merged the other PR. Noop strategy is also fine for me as well. I think the current state of the PR already looks good.

Updating the tests is ok. Release notes will have to go a bit into detail. But more than likely it simplifies things for all people.

The PagingBehaviorDirectory would maybe still worthwhile to have to move that logic out of the DefaultInformationProvider (there is many things in there already). But not a priority for this PR.

**Comment by weaselmetal** [comment]
Dang. service location through reflection on entire packages ain't cool.
VoidPagingBehavior, CustomOffsetLimitPagingBehavior (which is totally unrelated and irrelevant to the tests run afaik) and OffsetLimitPagingBehavior all are registered. By happenstance, the void behavior is then picked up (as it happens to be the first) and used a paging behavior on the resources.
Service location through reflection is going to lash back every single time an alternative (service) implementation is added and tests didn't expect that. Also, it (hopefully) never reflects how service location is set up in real projects. `ReflectionsServiceDiscovery` also hid the fact that setup of paging behavior through modules didn't work properly.
It may be convenient, but I thinks it's more a tool to shoot yourself in the foot.

**Comment by remmeier** [comment]
I don't mind sending that ReflectionServiceDiscovery to hell for that test. Once upon a time it was the default Katharsis mechanisms. Nowadays it is still supported by not recommended (and deprecated at various places). But would not mind to drop it entirely because there are plenty of more reasonable/lightweight alternatives.

**Comment by remmeier** [comment]
(but the dropping part would come with v3)

**Comment by weaselmetal** [comment]
Would be great @remmeier if you could check if a build of that state works with your real world projects.

**Comment by remmeier** [comment]
thx

**Pull request merged**